File: Java/nextGreaterElement.java
import java.util.*;

class Solution 
{
    public int[] findNextGreaterElement(int[] nums)
    {
        int[] result = new int[nums.length];
        Stack<Integer> stack = new Stack<>();
        for(int i=nums.length-1;i>=0;i--)
        {
            if(!stack.empty())
            {
                while(!stack.empty() && stack.peek() <= nums[i])
                {
                    stack.pop();
                }
            }
            result[i] = stack.empty() ? -1 : stack.peek();
            stack.push(nums[i]);
        }
        return result;
    }
}
public class nextGreaterElement 
{
    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int nums[] = new int[n];
        for(int i=0;i<n;i++) nums[i] = sc.nextInt();
        sc.close();

        Solution sol = new Solution();
        int[] result = sol.findNextGreaterElement(nums);
        for(int i=0;i<result.length;i++) System.out.print(result[i] + " ");
    }
}

File: C++/nextGreaterElement.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution 
{
    public:
    int checkGreaterElement(vector<int> &nums,int element,int index)
    {
        for(int i=index;i<nums.size();i++)
        {
            if(nums[i] > element) return nums[i];
        }
        return -1;
    }
    vector<int> nextGreaterElement(vector<int> &nums)
    {
        vector<int> result;
        for(int i=0;i<nums.size()-1;i++)
        {
            result.push_back(checkGreaterElement(nums,nums[i],i+1));
        }
        result.push_back(-1);
        return result;
    }
};
int main()
{
    int n;
    cin>>n;
    vector<int> nums(n);
    for(int i=0;i<n;i++) cin>>nums[i];
    Solution sol;
    vector<int> result = sol.nextGreaterElement(nums);
    for(auto i: result) cout<<i<<" ";
}
